docker-compose.yml

```diff
@@ -0,0 +1,210 @@
+    entrypoint: node
+    command:
+      - "/app/oppia/node_modules/webpack/bin/webpack.js"
+      - "--config"
+      - "/app/oppia/webpack.dev.config.ts"
+      - "--watch"
+      - "--color"
+      - "--progress"
+
+  google_app_engine:
+    build:
+      context: .
+      dockerfile: ./docker/Dockerfile
+    container_name: oppia-app-engine
+    ports:
+      - "8181:8181"
+      - "8000:8000"
+    volumes:
+      - .:/app/oppia
+      - shared_proto_files:/app/oppia/proto_files
+      - shared_third_party:/app/oppia/third_party
+      - shared_node_modules:/app/oppia/node_modules
+      - shared_js_ts_proto_files:/app/oppia/extensions/classifiers/proto
+      - shared_assets_dir:/app/oppia/assets
+      - shared_angular_build:/app/oppia/dist
+      - shared_webpack_bundles:/app/oppia/webpack_bundles
+    environment:
+      - CLOUDSDK_CORE_DISABLE_PROMPTS=1
+      - oppia_is_dockerized=true
+    depends_on:
+      angular-build:
+        condition: service_healthy
+      webpack-compiler:
+        condition: service_healthy
+    healthcheck:
+      test: curl -f http://localhost:8181/ || exit 1
+      timeout: 10s
+      interval: 30s
+      retries: 5
+      start_period: 20m
+    entrypoint: /google-cloud-sdk/bin/dev_appserver.py
+    command: 
+      - "/app/oppia/app_dev_docker.yaml"
+      - "--runtime"
+      - "python38"
+      - "--host=0.0.0.0"
+      - "--port=8181"
+      - "--admin_host=0.0.0.0"
+      - "--admin_port=8000"
+      - "--skip_sdk_update_check=True"
+
+  datastore:
+    build:
+      context: .
+      dockerfile: ./docker/Dockerfile
+    container_name: oppia-cloud-datastore
+    ports:
+      - "8089:8089"
+    environment:
+      - CLOUDSDK_CORE_DISABLE_PROMPTS=1
+    depends_on:
+      angular-build:
+        condition: service_healthy
+      webpack-compiler:
+        condition: service_healthy
+    healthcheck:
+      test: curl -f http://localhost:8089/ || exit 1
+      timeout: 10s
+      interval: 2m
+      retries: 5
+      start_period: 20m
+    entrypoint: /google-cloud-sdk/bin/gcloud
+    command:
+      - "beta"
+      - "emulators"
+      - "datastore"
+      - "start"
+      - "--project"
+      - "dev-project-id"
+      - "--data-dir"
+      - "/app/cloud_datastore_emulator_cache"
+      - "--host-port=0.0.0.0:8089"
+      - "--consistency=1.0"
+      - "--quiet"
+      - "--no-store-on-disk"
+
+  firebase:
+    image: node@sha256:48650ae5c5dc9fbd0d0559fd1248e779f7214ba99434420a089bc7853aeefb65
+    container_name: oppia-firebase-emulator
+    ports:
+      - "9099:9099"
+      - "4000:4000"
+    volumes:
+      - ./.firebase.json:/app/oppia/.firebase.json
+      - shared_node_modules:/app/oppia/node_modules
+    environment:
+      - oppia_is_dockerized=true
+    healthcheck:
+      test: wget --spider http://localhost:4000/auth/ && echo 'y' | /app/oppia/node_modules/firebase-tools/lib/bin/firebase.js emulators:export /app/firebase_emulator_cache --project=dev-project-id || exit 1
```

Comment:
Can you explain what are we doing here? both the 1st and 2nd command

Reply:
okay, so `wget --spider http://localhost:4000/auth/` is the actual healthcheck command (NOTE: wget --spider is an alternative for curl as this dockerhub image do not have curl downloaded)
another command is to export the auth users data along with each healthcheck execution

Reply:
> another command is to export the auth users data along with each healthcheck execution

Do we need to run it all the time? Do we do it right now when start the server using start.py?

Reply:
right now, it is executed as the firebase is stopped.

Reply:
In that case, why are we running it all the time, also how much time/resource does it take to execute?

Reply:
> right now, it is executed as the firebase server is stopped.

By `right now` do you mean when we are running the server using start.py? If yes, can you please share the link for ref. I wanted to check how we are running it at the end and can possible do it in the container?

Reply:
about resources, its the same as before it was not in the healthcheck!

yes, means current setup. this `--export-on-exit` flag ensures to export the auth data when server is exited
https://firebase.google.com/docs/emulator-suite/install_and_configure#startup